### PR TITLE
iso-image: Fix issues with UEFI installer image

### PR DESCRIPTION
**Pull request by samueldr**
Tries to handle #123376

###### Things done

Verified on:

 - qemu (-cdrom)
 - Acer c720p
 - Pinebook (A64)

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Added a release notes entry if the change is major or breaking
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/installer/cd-dvd/iso-image.nix [hunk 1]
```diff
@@ -403,7 +421,9 @@ let
       mkdir ./boot
       cp -p "${config.boot.kernelPackages.kernel}/${config.system.boot.loader.kernelFile}" \
         "${config.system.build.initialRamdisk}/${config.system.boot.loader.initrdFile}" ./boot/
-      touch --date=@0 ./EFI ./boot
+
+      # Rewrite dates for everything in the FS
+      find . -exec touch --date=@$DATE '{}' ';'
```
Comment (ncfavier) [inline nixos/modules/installer/cd-dvd/iso-image.nix:426]: You could just use `--date=2000-01-01` here to avoid magic numbers.

Comment (ncfavier) [inline nixos/modules/installer/cd-dvd/iso-image.nix:426]:
```suggestion
      find . -exec touch --date=2000-01-01 {} +
```

to process more than one file at once

**Comment by hyperfekt** [comment]
The date issue is fixed, the one of failed theming is not. I can't commit on the resolution one since I don't have that.

**Comment by hyperfekt** [comment]
@ncfavier Can you confirm the font problem is fixed? In that case we should merge a PR with either just the `(hd0) → ($root)` commit or that *and* the revert backed out and resolve the blocker.

**Comment by ncfavier** [comment]
Building the ISO now.

To be clear, `($root)/foo` is the same as just `/foo`. I don't think we should merge this until we figure out a way to properly set the `$root`, which probably involves using the `search` command.

**Comment by hyperfekt** [comment]
Well, that's just an issue of lacking theming, and the other commits prevent booting. Which is why, as I see it, the latter should prevent a release of 21.05 of while the former should not. This is why I'd like to separate the two parts and merge them separately, so we can take our time and find out how to solve the theming issue on all platforms without making @jonringer sad.

**Comment by ncfavier** [comment]
Fair enough.

**Comment by jonringer** [comment]
> without making @jonringer sad.

at this point, all software makes me sad

File: nixos/modules/installer/cd-dvd/iso-image.nix (hunk 1 above)
Comment (samueldr) [reply nixos/modules/installer/cd-dvd/iso-image.nix:426]: TBF a magic number is magic when used with neither name or a description :).

But yes, much preferable this way.

**Comment by ncfavier** [comment]
OK, I can confirm that this fixes the date bug and the resolution bug (at least on my machine).

d64034d5e79cf454a995d3270931d02c67287e0b and d64034d5e79cf454a995d3270931d02c67287e0b cancel each other out, so you might want to rebase them out.

For the record, [here's a list](https://en.wikipedia.org/wiki/VESA_BIOS_Extensions#Other_commonly_available_graphics_modes) of common VBE video modes.

**Comment by hyperfekt** [comment]
Fix is still good after the force-push.

**Comment by samueldr** [comment]
AFAIUI `vbe` is irrelevant here since we're in an UEFI gop driver.

**Comment by ncfavier** [comment]
From the manual:

> Note that you can only use modes which your graphics card supports via VESA BIOS Extensions (VBE)

I have no idea though.

**Comment by samueldr** [comment]
Yeah, GRUB's documentation is still heavily biased to their "legacy" program.

We're exclusively using the UEFI version of GRUB with the iso.

**Comment by ncfavier** [comment]
Aha you're right, `videoinfo` lists modes that aren't in that table.

**Comment by samueldr** [comment]
haha! detecting as `(hd0,msdos2)` breaks themes here! An artificial reproducer!

So I guess the issue was that GRUB detected itself as `(hd0,msdos2)`. (For the theme issue.)

**Comment by hyperfekt** [comment]
The latest commit is unsuccessful in fixing the theming issue for me, in case that is what it is supposed to do.

**Comment by samueldr** [comment]
@hyperfekt yeah, I pushed it eagerly, and then I figured out that reproduces the themeing issue.

**Comment by samueldr** [comment]
I need to test on `AArch64` still.

**Comment by hyperfekt** [comment]
This is the one. Theming works now, and the date thing is still resolved. Good job!

**Comment by jonringer** [comment]
awesome turn around. Thanks for working on this all :)

**Comment by samueldr** [comment]
Verified with *[Tow-Boot](https://tow-boot.org/)* on my Pinebook A64 that it works just as well as it did.

**Comment by jonringer** [comment]
@GrahamcOfBorg test grub

**Review by jonringer — APPROVED** [review]
diff LGTM

tests still checkout on x86, aarch64 was tested

**Comment by ncfavier** [comment]
This works for me, but only by accident: the `nixos-installer-image` marker file ends up on both `(hd0)` and `(hd0,msdos2)`, so whichever GRUB tries first is picked. We should make sure to use a file that's only on `(hd0)`.

**Comment by samueldr** [comment]
It's not actually a different FS. It's `isohybrid` trickery.

We should look into why some files are not present.

But for the time being the failure state is seemingly lack of theme. Which isn't much of an issue as long as the user can boot in the system!

**Comment by codelamb** [comment]
Hi,  

I have noticed that  iso-image plasma5-20.09.2538.0cfd08f4881-x86_64-linux.iso  probably lacks tools for dealing with fat32 filesystems. I had issues installing Nixos 21.05 on Samsung NP530U4E. System was damaged while upgrading from 20.03 to 21.05, giving the error "unable to build, disc was probably full" (quoting from memory, sorry). After trying to reinstall from USB got the same error message. Gparted was showing error message, that system lacks tools to deal with fat32 filesystems.I assume that  nixos-install was probably unable to detect fat32 boot partition and was assuming that disk is full. I was strictly following [official manual for UEFI systems](https://nixos.org/manual/nixos/stable/index.html#sec-installation-partitioning-UEFI). I was able to install Nixos only after manual installation of mtools with `nix-env -f '<nixpkgs>' -iA mtools`. Just in case I have also declared mtools in configuration.nix

**Comment by samueldr** [comment]
(I "moved" this comment into a new issue as it is irrelevant to this PR see #132094 )

**Pull request merged**